tempodb/encoding/v2/data_reader.go

```diff
@@ -23,41 +27,97 @@ var constDataHeader = &dataHeader{}
 func NewDataReader(r backend.ContextReader, encoding backend.Encoding) (common.DataReader, error) {
-	v2DataReader := &dataReader{
-		contextReader: r,
-		dataReader:    v0.NewDataReader(r),
-	}
-
-	// wrap the paged reader in a compressed/v1 reader and return that
-	v1DataReader, err := v1.NewNestedDataReader(v2DataReader, encoding)
+	pool, err := getReaderPool(encoding)
 	if err != nil {
 		return nil, err
 	}
 
-	return v1DataReader, nil
+	return &dataReader{
+		contextReader: r,
+		pool:          pool,
+	}, nil
 }
 
 // Read implements common.DataReader
 func (r *dataReader) Read(ctx context.Context, records []common.Record, pagesBuffer [][]byte, buffer []byte) ([][]byte, []byte, error) {
-	pagesBuffer, buffer, err := r.dataReader.Read(ctx, records, pagesBuffer, buffer)
+	if len(records) == 0 {
+		return nil, buffer, nil
+	}
+
+	start := records[0].Start
+	length := uint32(0)
+	for _, record := range records {
+		length += record.Length
+	}
+
+	if cap(buffer) < int(length) {
+		buffer = make([]byte, length)
+	}
+	buffer = buffer[:length]
+	_, err := r.contextReader.ReadAt(ctx, buffer, int64(start))
 	if err != nil {
 		return nil, nil, err
 	}
 
-	for i := range pagesBuffer {
-		p, err := unmarshalPageFromBytes(pagesBuffer[i], constDataHeader)
+	compressedPagesBuffer := make([][]byte, len(records))
+
+	cursor := uint32(0)
+	previousEnd := uint64(0)
+	for i, record := range records {
+		end := cursor + record.Length
+		if end > uint32(len(buffer)) {
+			return nil, nil, fmt.Errorf("record out of bounds while reading pages: %d, %d, %d, %d", cursor, record.Length, end, len(buffer))
+		}
+
+		if previousEnd != record.Start && previousEnd != 0 {
+			return nil, nil, fmt.Errorf("non-contiguous pages requested from dataReader: %d, %+v", previousEnd, record)
+		}
+
+		compressedPagesBuffer[i] = buffer[cursor:end]
+		cursor += record.Length
+		previousEnd = record.Start + uint64(record.Length)
+	}
+
+	// read and strip page data
+	compressedPages := make([][]byte, 0, len(compressedPagesBuffer))
+	for _, v0Page := range compressedPagesBuffer {
+		page, err := unmarshalPageFromBytes(v0Page, constDataHeader)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		compressedPages = append(compressedPages, page.data)
+	}
+
+	// prepare pagesBuffer
+	if cap(pagesBuffer) < len(compressedPages) {
+		// extend pagesBuffer
+		diff := len(compressedPages) - cap(pagesBuffer)
+		pagesBuffer = append(pagesBuffer[:cap(pagesBuffer)], make([][]byte, diff)...)
+	} else {
+		pagesBuffer = pagesBuffer[:len(compressedPages)]
+	}
+
+	// now decompress
+	for i, page := range compressedPages {
+		reader, err := r.getCompressedReader(page)
 		if err != nil {
 			return nil, nil, err
 		}
 
-		pagesBuffer[i] = p.data
+		pagesBuffer[i], err = tempo_io.ReadAllWithBuffer(reader, len(page), pagesBuffer[i])
+		if err != nil {
+			return nil, nil, err
+		}
 	}
 
 	return pagesBuffer, buffer, nil
 }
 
 func (r *dataReader) Close() {
-	r.dataReader.Close()
```

Comment:
Do we still need to do something here to close this out?

Reply:
Nope, b/c there is no longer an encapsulated data reader.